src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultHtmlGenerator.cs

```diff
@@ -746,18 +725,130 @@ public virtual TagBuilder GenerateAntiForgery([NotNull] ViewContext viewContext)
         }
 
+        /// <inheritdoc />
+        public virtual IReadOnlyCollection<string> GetCurrentValues(
+            [NotNull] ViewContext viewContext,
+            ModelExplorer modelExplorer,
+            string expression,
+            bool allowMultiple)
+        {
+            var fullName = GetFullHtmlFieldName(viewContext, expression);
+            if (string.IsNullOrEmpty(fullName))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(expression));
+            }
+
+            var type = allowMultiple ? typeof(string[]) : typeof(string);
+            var rawValue = GetModelStateValue(viewContext, fullName, type);
+
+            // If ModelState did not contain a current value, fall back to ViewData- or ModelExplorer-supplied value.
+            if (rawValue == null)
+            {
+                if (modelExplorer == null)
+                {
+                    // Html.DropDownList() and Html.ListBox() helper case.
+                    rawValue = viewContext.ViewData.Eval(expression);
+                    if (rawValue is IEnumerable<SelectListItem>)
+                    {
+                        // This ViewData item contains the fallback selectList collection for GenerateSelect().
+                        // Do not try to use this collection.
+                        rawValue = null;
+                    }
+                }
+                else
+                {
+                    // <select/>, Html.DropDownListFor() and Html.ListBoxFor() helper case. Do not use ViewData.
+                    rawValue = modelExplorer.Model;
+                }
+
+                if (rawValue == null)
+                {
+                    return null;
+                }
+            }
+
+            // Convert raw value to a collection.
+            IEnumerable rawValues;
+            if (allowMultiple)
+            {
+                rawValues = rawValue as IEnumerable;
+                if (rawValues == null || rawValues is string)
+                {
+                    throw new InvalidOperationException(
+                        Resources.FormatHtmlHelper_SelectExpressionNotEnumerable(nameof(expression)));
+                }
+            }
+            else
+            {
+                rawValues = new[] { rawValue };
+            }
+
+            modelExplorer = modelExplorer ??
+                ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
+
+            var enumNames = modelExplorer.Metadata.EnumNamesAndValues;
+            var isEnum = modelExplorer.Metadata.IsEnum;
+            var innerType =
+                Nullable.GetUnderlyingType(modelExplorer.Metadata.ModelType) ?? modelExplorer.Metadata.ModelType;
+
+            // Convert raw value collection to strings.
+            var currentValues = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            foreach (var value in rawValues)
```

Comment:
loop got simpler once it didn't have to generate just one `HashSet` entry per `value`